geode-assembly/src/distributedTest/java/org/apache/geode/rest/internal/web/controllers/RestAPICompatibilityTest.java

```diff
@@ -0,0 +1,196 @@
+    int locatorPort2 = locator2.getPort();
+    cluster
+        .startServerVM(2, oldVersion, s -> s.withRegion(RegionShortcut.PARTITION, "region")
+            .withConnectionToLocator(locatorPort1, locatorPort2));
+    cluster
+        .startServerVM(3, oldVersion, s -> s.withRegion(RegionShortcut.PARTITION, "region")
+            .withConnectionToLocator(locatorPort1, locatorPort2));
+
+    // Roll locators to the current version
+    cluster.stop(0);
+    // gradle sets a property telling us where the build is located
+    final String buildDir = System.getProperty("geode.build.dir", System.getProperty("user.dir"));
+    locator1 = cluster.startLocatorVM(0, l -> l.withHttpService().withPort(locatorPort1)
+        .withConnectionToLocator(locatorPort2)
+        .withSystemProperty("geode.build.dir", buildDir));
+    cluster.stop(1);
+
+    cluster.startLocatorVM(1,
+        x -> x.withConnectionToLocator(locatorPort1).withHttpService().withPort(locatorPort2)
+            .withConnectionToLocator(locatorPort1)
+            .withSystemProperty("geode.build.dir", buildDir));
+
+    gfsh.connectAndVerify(locator1);
+    gfsh.execute("list members");
+    // Execute REST api calls to from the new locators to the old servers to ensure that backwards
+    // compatibility is maintained
+
+    executeAndValidatePOSTRESTCalls(locator1.getHttpPort());
+    executeAndValidateGETRESTCalls(locator1.getHttpPort());
+
+  }
+
+  void executeAndValidatePOSTRESTCalls(int locator) throws Exception {
+
+    CloseableHttpClient httpClient = HttpClients.createDefault();
+    for (Map.Entry<String, String[]> entry : postRESTAPICalls.entrySet()) {
+      // Skip the test is the version is before the REST api was introduced.
+      if (TestVersion.compare(oldVersion, entry.getValue()[2]) < 0) {
+        continue;
+      }
+      HttpPost post =
+          new HttpPost("http://localhost:" + locator + entry.getKey());
+      post.addHeader("Content-Type", "application/json");
+      post.addHeader("Accept", "application/json");
+      StringEntity jsonStringEntity =
+          new StringEntity(entry.getValue()[0], ContentType.DEFAULT_TEXT);
+      post.setEntity(jsonStringEntity);
+      CloseableHttpResponse response = httpClient.execute(post);
+
+      HttpEntity entity = response.getEntity();
+      InputStream content = entity.getContent();
+
+      BufferedReader reader = new BufferedReader(new InputStreamReader(content));
+      String line;
+      StringBuilder sb = new StringBuilder();
+      while ((line = reader.readLine()) != null) {
+        sb.append(line);
+      }
+      JsonNode jsonObject = mapper.readTree(sb.toString());
+      String statusCode = jsonObject.findValue("statusCode").textValue();
+      assertTrue(statusCode.equals("ACCEPTED") || statusCode.contains("OK"));
+      String statusMessage = jsonObject.findValue("statusMessage").textValue();
+      assertTrue(statusMessage.contains(entry.getValue()[1]));
```

Comment:
This assertion will also just fail with:
```
java.lang.AssertionError 
        at org.junit.Assert.fail(Assert.java:87) 
        at org.junit.Assert.assertTrue(Assert.java:42) 
        at org.junit.Assert.assertTrue(Assert.java:53)
	at org.apache.geode.rest.internal.web.controllers.RestAPICompatibilityTest.executeAndValidatePOSTRESTCalls(RestAPICompatibilityTest.java:170)
	at org.apache.geode.rest.internal.web.controllers.RestAPICompatibilityTest.restCommandExecutedOnLatestLocatorShouldBeBackwardsCompatible (RestAPICompatibilityTest.java:135)
```
While AssertJ failure will look like:
```
java.lang.AssertionError: 
Expecting:
  "A really long message that has other entry value"
to contain:
  "expected entry value" 

	at org.apache.geode.rest.internal.web.controllers.RestAPICompatibilityTest.executeAndValidatePOSTRESTCalls(RestAPICompatibilityTest.java:170)
	at org.apache.geode.rest.internal.web.controllers.RestAPICompatibilityTest.restCommandExecutedOnLatestLocatorShouldBeBackwardsCompatible (RestAPICompatibilityTest.java:135)
```